### PR TITLE
Improve docs clarity on loading custom checkers - FIXED

### LINKED ISSUE #7264: Cannot get pylint to register custom checker, is there a missing step in the docs?
### Question

Hi! First: Thanks for supporting and maintaining such a great ecosystem!

My question: I'm trying to make a simple checker around some custom formatting in log strings. I have followed every tutorial I can find from within the last few years, and naturally looked at the guide [in the documentation](https://pylint.pycqa.org/en/v2.14.5/development_guide/how_tos/custom_checkers.html).

However, I am finding that the `register` function, described as the way to get your checker actually loaded into pylint, simply is not being called. The module is being loaded, for sure — I can see breakpoints in the top level of the module being hit, but not in the `register` function itself. I've tried setting the `--load-plugins` argument, I've tried using the `load_plugins` setting in pylintrc.

Is there an additional step or constraint missing in the implementation of custom checkers from the docs? Or am I simply making a mistake?

Below is the code for my absolute bare-bones first attempt.

Stored in `testchecker.py`:
```python
from astroid import nodes
from pylint.checkers import BaseChecker

breakpoint() # HIT!

class SimpleCheck(BaseChecker):
    """Simple"""
    name = "simple-test-check"
    msgs = {
            "W6900": ("Test", "test-check", "Just a test")
            }
    options = ( ("fake-option", {
                "default": False,
                "type": "yn", # EDIT: bool here was not allowed
                "metavar": "<y or n>",
                "help": "Some fake option"}
                ),
            )

    def visit_import(self, _: nodes.Import) -> None:
        breakpoint() # NOT HIT!

def register(linter):
    breakpoint() # NOT HIT!
    linter.register_checker(SimpleCheck(linter))
```

My file I am linting, `lintme.py`:
```python
"""MODULE DOCSTRING"""
import logging

logger = logging.getLogger(__name__)

logger.info("hello world")
```
Both files are in the same directory, in which I execute:
```bash
pylint --load-plugins=testchecker lintme.py
```

```
$ pylint --version
pylint 2.14.5
astroid 2.11.7
Python 3.10.4 (main, May 19 2022, 08:51:27) [Clang 13.1.6 (clang-1316.0.21.2.5)]
$ uname -a
Darwin OVO4941L.ovoenergy.local 21.6.0 Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:25 PDT 2022; root:xnu-8020.140.41~1/RELEASE_X86_64 x86_64
```

Thanks again!

### Documentation for future user

I expect the guide to creating a custom checker to contain everything needed to do so, and to have a section highlighting common and easy mistakes (I'm assuming I have made such a mistake)

### Additional context

I have searched stack overflow and pre-existing issues on this repo, and not found anyone reporting this.
There is https://stackoverflow.com/questions/71590677/cant-load-custom-pylint-module, but they report getting an error about not finding the module, then succeeding by placing the module in the pylint source. This is not what I am experiencing.

### PR DESCRIPTION
**Pull request by daogilvie**
NOTE: This PR replaces #7276 due to an admin error on my part.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :scroll: Docs          |

### Short
This PR just updates the documentation page on writing custom checkers. It adds
information about some restrictions between how `init-hook` changes the loading
path and when you can use that to import your custom module.
### Long (in the commit message)
Whilst the docs on testing custom checkers do touch on the fact the
module has to be in your load path, there are some restrictions that
were not made clear before.
Specifically, the init-hook configuration item (which is often used
explicitly to change the sys.path) is not automatically sufficient. If
the init hook is in a file, but the load-plugins argument is passed in
the CLI, then the plugins are loaded before the init-hook has run, and
so will not be imported. In this case, the failure can also be silent,
so the outcome is a lint that will simply not contain any information
from the checker, as opposed to a run that will error.
This corner case may merit fixing in a separate PR, that is to be
confirmed.

Closes #7264

### CODE REVIEW

File: doc/development_guide/how_tos/custom_checkers.rst [hunk 1]
```diff
@@ -218,10 +218,29 @@ Now we can debug our checker!
...
+    If your pylint config has an ``init-hook`` that modifies
+    ``sys.path`` to include the module's parent directory, this
+    will also work, but only if either:
+
+    * the ``init-hook`` and the ``load-plugins`` list are both
+      defined in a configuration file, or...
+    * the ``init-hook`` is passed as a command-line argument and
+      the ``load-plugins`` list is in the configuration file
+
+    I.E You cannot load a custom plugin by modifying ``sys.path`` if you
```
Comment (DanielNoord) [inline doc/development_guide/how_tos/custom_checkers.rst:238]:
```suggestion
    So, you cannot load a custom plugin by modifying ``sys.path`` if you
```

File: doc/development_guide/how_tos/custom_checkers.rst [hunk 2]
```diff
@@ -218,10 +218,29 @@ Now we can debug our checker!
...
+    supply the ``init-hook`` in a configuration file, but pass the module name
+    in via ``--load-plugins`` on the command line.
+
```
Comment (DanielNoord) [inline doc/development_guide/how_tos/custom_checkers.rst:241]:
```suggestion
```

File: doc/whatsnew/fragments/7264.other [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Dev documentation is now clearer about custom checker loading.
```
Comment (DanielNoord) [inline doc/whatsnew/fragments/7264.other:1]: I think this can be removed.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Final points!

**Review by DanielNoord — APPROVED** [review]

**Comment by daogilvie** [comment]
You are very welcome, thanks for improving and maintaining a project that has saved me from myself more times than I can remember 😂

**Pull request merged**